primitives/src/conststr.rs

```diff
@@ -453,6 +469,138 @@ impl TryFrom<String> for InlinedString {
     }
 }
 
+mod ffi {
+    #![allow(unsafe_code)]
+    use core::mem::ManuallyDrop;
+
+    use iroha_ffi::{
+        ir::{Ir, Transmute},
+        repr_c::{COutPtr, CType, CTypeConvert},
+        slice::OutBoxedSlice,
+    };
+
+    use super::*;
+
+    #[repr(C)]
+    #[derive(Clone, Copy)]
+    pub union ReprCConstString {
+        inlined: InlinedString,
+        boxed: ReprCBoxedString,
+    }
+
+    #[repr(C)]
+    #[derive(Debug, Clone, Copy)]
+    pub struct ReprCBoxedString {
+        #[cfg(target_endian = "little")]
+        ptr: *mut u8,
+        len: usize,
+        #[cfg(target_endian = "big")]
+        ptr: *mut u8,
+    }
+
+    impl ReprCConstString {
+        #[inline]
+        const unsafe fn is_inlined(&self) -> bool {
+            // SAFETY: access to the MSB is always safe regardless of the correct variant.
+            self.inlined.is_inlined()
+        }
+    }
+
+    iroha_ffi::ffi_type! {unsafe impl Robust for ReprCConstString}
+
+    unsafe impl Transmute for &ConstString {
+        type Target = *const ReprCConstString;
+
+        unsafe fn is_valid(target: &Self::Target) -> bool {
+            !target.is_null()
+        }
+    }
+
+    impl Ir for ConstString {
+        type Type = Self;
+    }
+
+    impl CType for ConstString {
+        type ReprC = ReprCConstString;
+    }
+    impl CTypeConvert<'_, ReprCConstString> for ConstString {
+        type RustStore = Self;
+        type FfiStore = ();
+
+        fn into_repr_c(self, store: &mut Self::RustStore) -> ReprCConstString {
+            *store = self;
+
+            if store.is_inlined() {
+                ReprCConstString {
+                    inlined: *store.inlined(),
+                }
+            } else {
+                let boxed = unsafe {
+                    ReprCBoxedString {
+                        ptr: store.boxed.ptr.as_ptr(),
+                        len: store.boxed.len,
+                    }
+                };
+
+                ReprCConstString { boxed }
+            }
+        }
+
+        unsafe fn try_from_repr_c(source: ReprCConstString, _: &mut ()) -> iroha_ffi::Result<Self> {
+            if source.is_inlined() {
+                return Ok(ConstString {
+                    inlined: source.inlined,
+                });
+            }
+
+            let boxed = ManuallyDrop::new(BoxedString {
+                ptr: core::ptr::NonNull::new_unchecked(source.boxed.ptr),
```

Comment:
Do we have any gurantee that recieved through ffi pointer would be non-null?

Reply:
yes, good catch